File: app/rules.py
"""Categorization rule preview and creation."""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal

import psycopg
from psycopg import Connection

from app.classify.pipeline import (
    CategorizationRule,
    ClassificationFlags,
    MatchField,
    MatchOperator,
    TransactionForClassification,
    rule_matches,
)

RULE_FIELDS = ("description", "counterparty_name", "counterparty_iban", "merchant")
RULE_OPERATORS = ("contains", "exact", "regex", "starts_with", "ends_with")


@dataclass(frozen=True)
class RuleDraft:
    source_transaction_id: int | None
    name: str
    field: str
    operator: str
    pattern: str
    category_name: str
    merchant_name: str | None


@dataclass(frozen=True)
class RulePreview:
    draft: RuleDraft
    match_count: int
    would_change_count: int
    already_correct_count: int
    manual_overrides_skipped_count: int


@dataclass(frozen=True)
class StoredRule:
    id: int
    name: str
    priority: int
    is_active: bool
    field: str
    operator: str
    pattern: str
    category_name: str | None
    merchant_name: str | None


@dataclass(frozen=True)
class RuleHistoryPreview:
    match_count: int
    would_change_count: int
    already_correct_count: int
    manual_overrides_skipped_count: int


@dataclass(frozen=True)
class RuleListItem:
    rule: StoredRule
    preview: RuleHistoryPreview


@dataclass(frozen=True)
class RuleBackfillResult:
    updated_count: int
    skipped_manual_count: int


@dataclass(frozen=True)
class RuleEditorInput:
    name: str
    priority: int
    is_active: bool
    field: str
    operator: str
    pattern: str
    category_name: str
    merchant_name: str | None


def draft_rule_from_transaction(database_url: str, transaction_id: int) -> RuleDraft:
    with psycopg.connect(database_url) as connection:
        row = connection.execute(
            """
            SELECT
                raw_transactions.counterparty_name,
                raw_transactions.description,
                categories.name,
                merchants.name
            FROM enriched_transactions
            JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
            LEFT JOIN categories ON categories.id = enriched_transactions.category_id
            LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
            WHERE enriched_transactions.id = %s
            """,
            (transaction_id,),
        ).fetchone()
    if row is None:
        raise ValueError(f"transaction not found: {transaction_id}")

    counterparty = _optional_str(row[0])
    description = str(row[1])
    pattern = counterparty or description
    field = "counterparty_name" if counterparty else "description"
    return RuleDraft(
        source_transaction_id=transaction_id,
        name=f"Classify {pattern}",
        field=field,
        operator="contains",
        pattern=pattern,
        category_name=str(row[2] or "Unknown"),
        merchant_name=_optional_str(row[3]),
    )


def preview_rule(database_url: str, draft: RuleDraft) -> RulePreview:
    with psycopg.connect(database_url) as connection:
        rows = connection.execute(
            """
            SELECT
                categories.name,
                manual_overrides.id IS NOT NULL
            FROM enriched_transactions
            JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
            LEFT JOIN categories ON categories.id = enriched_transactions.category_id
            LEFT JOIN manual_overrides
                ON manual_overrides.enriched_transaction_id = enriched_transactions.id
            WHERE
                CASE
                    WHEN %s = 'counterparty_name'
                        THEN raw_transactions.counterparty_name ILIKE '%%' || %s || '%%'
                    ELSE raw_transactions.description ILIKE '%%' || %s || '%%'
                END
            """,
            (draft.field, draft.pattern, draft.pattern),
        ).fetchall()

    manual_skipped = sum(1 for row in rows if row[1])
    eligible_rows = [row for row in rows if not row[1]]
    already_correct = sum(1 for row in eligible_rows if row[0] == draft.category_name)
    would_change = len(eligible_rows) - already_correct

    return RulePreview(
        draft=draft,
        match_count=len(rows),
        would_change_count=would_change,
        already_correct_count=already_correct,
        manual_overrides_skipped_count=manual_skipped,
    )


def create_rule(database_url: str, draft: RuleDraft) -> int:
    with psycopg.connect(database_url) as connection:
        with connection.transaction():
            merchant_id = _merchant_id(connection, draft.merchant_name)
            row = connection.execute(
                """
                INSERT INTO categorization_rules (
                    name,
                    priority,
                    field,
                    operator,
                    pattern,
                    category_id,
                    merchant_id,
                    created_from_transaction_id
                )
                VALUES (
                    %s,
                    100,
                    %s,
                    %s,
                    %s,
                    (SELECT id FROM categories WHERE name = %s),
                    %s,
                    %s
                )
                RETURNING id
                """,
                (
                    draft.name,
                    draft.field,
                    draft.operator,
                    draft.pattern,
                    draft.category_name,
                    merchant_id,
                    draft.source_transaction_id,
                ),
            ).fetchone()
    if row is None:
        raise RuntimeError("rule insert did not return an id")
    return _read_int(row[0])


def create_rule_from_input(database_url: str, rule_input: RuleEditorInput) -> int:
    with psycopg.connect(database_url) as connection:
        with connection.transaction():
            category_id = _category_id(connection, rule_input.category_name)
            merchant_id = _upsert_merchant_id(connection, rule_input.merchant_name, category_id)
            row = connection.execute(
                """
                INSERT INTO categorization_rules (
                    name,
                    priority,
                    is_active,
                    field,
                    operator,
                    pattern,
                    category_id,
                    merchant_id
                )
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                RETURNING id
                """,
                (
                    rule_input.name,
                    rule_input.priority,
                    rule_input.is_active,
                    rule_input.field,
                    rule_input.operator,
                    rule_input.pattern,
                    category_id,
                    merchant_id,
                ),
            ).fetchone()
    if row is None:
        raise RuntimeError("rule insert did not return an id")
    return _read_int(row[0])


def update_rule_from_input(database_url: str, rule_id: int, rule_input: RuleEditorInput) -> None:
    with psycopg.connect(database_url) as connection:
        with connection.transaction():
            category_id = _category_id(connection, rule_input.category_name)
            merchant_id = _upsert_merchant_id(connection, rule_input.merchant_name, category_id)
            connection.execute(
                """
                UPDATE categorization_rules
                SET
                    name = %s,
                    priority = %s,
                    is_active = %s,
                    field = %s,
                    operator = %s,
                    pattern = %s,
                    category_id = %s,
                    merchant_id = %s,
                    updated_at = now()
                WHERE id = %s
                """,
                (
                    rule_input.name,
                    rule_input.priority,
                    rule_input.is_active,
                    rule_input.field,
                    rule_input.operator,
                    rule_input.pattern,
                    category_id,
                    merchant_id,
                    rule_id,
                ),
            )


def parse_rule_editor_input(form: dict[str, str]) -> RuleEditorInput:
    name = form.get("name", "").strip()
    field = form.get("field", "").strip()
    operator = form.get("operator", "").strip()
    pattern = form.get("pattern", "").strip()
    category_name = form.get("category", "").strip()
    merchant_name = form.get("merchant", "").strip() or None
    if not name:
        raise ValueError("rule name is required")
    if field not in RULE_FIELDS:
        raise ValueError(f"unsupported rule field: {field}")
    if operator not in RULE_OPERATORS:
        raise ValueError(f"unsupported rule operator: {operator}")
    if not pattern:
        raise ValueError("rule pattern is required")
    if not category_name:
        raise ValueError("rule category is required")
    return RuleEditorInput(
        name=name,
        priority=_parse_priority(form.get("priority", "")),
        is_active=form.get("is_active") == "1",
        field=field,
        operator=operator,
        pattern=pattern,
        category_name=category_name,
        merchant_name=merchant_name,
    )


def list_rules(database_url: str) -> list[RuleListItem]:
    with psycopg.connect(database_url) as connection:
        rules = _load_stored_rules(connection)
        return [RuleListItem(rule=rule, preview=_preview_stored_rule(connection, rule.id)) for rule in rules]


def set_rule_active(database_url: str, rule_id: int, *, is_active: bool) -> None:
    with psycopg.connect(database_url) as connection:
        with connection.transaction():
            connection.execute(
                "UPDATE categorization_rules SET is_active = %s, updated_at = now() WHERE id = %s",
                (is_active, rule_id),
            )


def backfill_rule(database_url: str, rule_id: int) -> RuleBackfillResult:
    with psycopg.connect(database_url) as connection:
        with connection.transaction():
            rule = _load_rule_for_matching(connection, rule_id)
            if rule is None:
                raise ValueError(f"rule not found: {rule_id}")
            if not rule.is_active:
                return RuleBackfillResult(updated_count=0, skipped_manual_count=0)

            matches = _matching_transactions(connection, rule)
            updated_count = 0
            skipped_manual_count = 0
            for match in matches:
                if match.has_manual_override:
                    skipped_manual_count += 1
                    continue
                _apply_rule_to_transaction(connection, rule, match.transaction_id)
                updated_count += 1

    return RuleBackfillResult(updated_count=updated_count, skipped_manual_count=skipped_manual_count)


def _load_stored_rules(connection: Connection[tuple[object, ...]]) -> list[StoredRule]:
    rows = connection.execute(
        """
        SELECT
            categorization_rules.id,
            categorization_rules.name,
            categorization_rules.priority,
            categorization_rules.is_active,
            categorization_rules.field,
            categorization_rules.operator,
            categorization_rules.pattern,
            categories.name,
            merchants.name
        FROM categorization_rules
        LEFT JOIN categories ON categories.id = categorization_rules.category_id
        LEFT JOIN merchants ON merchants.id = categorization_rules.merchant_id
        ORDER BY categorization_rules.priority, categorization_rules.id
        """
    ).fetchall()
    return [
        StoredRule(
            id=_read_int(row[0]),
            name=str(row[1]),
            priority=_read_int(row[2]),
            is_active=bool(row[3]),
            field=str(row[4]),
            operator=str(row[5]),
            pattern=str(row[6]),
            category_name=_optional_str(row[7]),
            merchant_name=_optional_str(row[8]),
        )
        for row in rows
    ]


@dataclass(frozen=True)
class _RuleTransactionMatch:
    transaction_id: int
    category_name: str | None
    merchant_name: str | None
    has_manual_override: bool


def _preview_stored_rule(connection: Connection[tuple[object, ...]], rule_id: int) -> RuleHistoryPreview:
    rule = _load_rule_for_matching(connection, rule_id)
    if rule is None:
        return RuleHistoryPreview(0, 0, 0, 0)
    matches = _matching_transactions(connection, rule)
    skipped = sum(1 for match in matches if match.has_manual_override)
    eligible = [match for match in matches if not match.has_manual_override]
    already_correct = sum(
        1
        for match in eligible
        if match.category_name == rule.category and (rule.merchant is None or match.merchant_name == rule.merchant)
    )
    return RuleHistoryPreview(
        match_count=len(matches),
        would_change_count=len(eligible) - already_correct,
        already_correct_count=already_correct,
        manual_overrides_skipped_count=skipped,
    )


def _load_rule_for_matching(connection: Connection[tuple[object, ...]], rule_id: int) -> CategorizationRule | None:
    row = connection.execute(
        """
        SELECT
            categorization_rules.id,
            categorization_rules.name,
            categorization_rules.priority,
            categorization_rules.field,
            categorization_rules.operator,
            categorization_rules.pattern,
            categories.name,
            merchants.name,
            categorization_rules.set_is_income,
            categorization_rules.set_is_transfer,
            categorization_rules.set_is_savings,
            categorization_rules.set_is_fixed_cost,
            categorization_rules.set_is_excluded_from_budget,
            categorization_rules.is_active
        FROM categorization_rules
        LEFT JOIN categories ON categories.id = categorization_rules.category_id
        LEFT JOIN merchants ON merchants.id = categorization_rules.merchant_id
        WHERE categorization_rules.id = %s
        """,
        (rule_id,),
    ).fetchone()
    if row is None:
        return None
    return CategorizationRule(
        id=_read_int(row[0]),
        name=str(row[1]),
        priority=_read_int(row[2]),
        field=MatchField(str(row[3])),
        operator=MatchOperator(str(row[4])),
        pattern=str(row[5]),
        category=_optional_str(row[6]),
        merchant=_optional_str(row[7]),
        flags=ClassificationFlags(
            is_income=_optional_bool(row[8]),
            is_transfer=_optional_bool(row[9]),
            is_savings=_optional_bool(row[10]),
            is_fixed_cost=_optional_bool(row[11]),
            is_excluded_from_budget=_optional_bool(row[12]),
        ),
        is_active=bool(row[13]),
    )


def _matching_transactions(
    connection: Connection[tuple[object, ...]],
    rule: CategorizationRule,
) -> list[_RuleTransactionMatch]:
    rows = connection.execute(
        """
        SELECT
            enriched_transactions.id,
            raw_transactions.account_id,
            raw_transactions.amount,
            raw_transactions.description,
            raw_transactions.counterparty_name,
            raw_transactions.counterparty_iban,
            merchants.name,
            categories.name,
            manual_overrides.id IS NOT NULL
        FROM enriched_transactions
        JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
        LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
        LEFT JOIN categories ON categories.id = enriched_transactions.category_id
        LEFT JOIN manual_overrides ON manual_overrides.enriched_transaction_id = enriched_transactions.id
        ORDER BY enriched_transactions.id
        """
    ).fetchall()
    matches: list[_RuleTransactionMatch] = []
    for row in rows:
        transaction = TransactionForClassification(
            id=_read_int(row[0]),
            account_id=_read_int(row[1]),
            amount=_read_decimal(row[2]),
            description=str(row[3]),
            counterparty_name=_optional_str(row[4]),
            counterparty_iban=_optional_str(row[5]),
            merchant_name=_optional_str(row[6]),
        )
        if rule_matches(transaction, rule):
            matches.append(
                _RuleTransactionMatch(
                    transaction_id=transaction.id,
                    category_name=_optional_str(row[7]),
                    merchant_name=transaction.merchant_name,
                    has_manual_override=bool(row[8]),
                )
            )
    return matches


def _apply_rule_to_transaction(
    connection: Connection[tuple[object, ...]],
    rule: CategorizationRule,
    transaction_id: int,
) -> None:
    connection.execute(
        """
        UPDATE enriched_transactions
        SET
            category_id = COALESCE((SELECT id FROM categories WHERE name = %s), category_id),
            merchant_id = COALESCE((SELECT id FROM merchants WHERE name = %s), merchant_id),
            is_income = COALESCE(%s, is_income),
            is_transfer = COALESCE(%s, is_transfer),
            is_savings = COALESCE(%s, is_savings),
            is_fixed_cost = COALESCE(%s, is_fixed_cost),
            is_excluded_from_budget = COALESCE(%s, is_excluded_from_budget),
            needs_review = false,
            classification_method = 'rule',
            classification_confidence = 1,
            classification_reason = %s,
            classification_model = NULL,
            classification_prompt_version = NULL,
            rule_id = %s,
            updated_at = now()
        WHERE id = %s
            AND NOT EXISTS (
                SELECT 1
                FROM manual_overrides
                WHERE manual_overrides.enriched_transaction_id = enriched_transactions.id
            )
        """,
        (
            rule.category,
            rule.merchant,
            rule.flags.is_income,
            rule.flags.is_transfer,
            rule.flags.is_savings,
            rule.flags.is_fixed_cost,
            rule.flags.is_excluded_from_budget,
            f"Backfilled from rule: {rule.name}",
            rule.id,
            transaction_id,
        ),
    )


def _merchant_id(connection: Connection[tuple[object, ...]], merchant_name: str | None) -> int | None:
    if merchant_name is None:
        return None
    row = connection.execute("SELECT id FROM merchants WHERE name = %s", (merchant_name,)).fetchone()
    if row is None:
        return None
    return _read_int(row[0])


def _upsert_merchant_id(
    connection: Connection[tuple[object, ...]],
    merchant_name: str | None,
    category_id: int,
) -> int | None:
    if merchant_name is None or merchant_name.strip() == "":
        return None
    normalized_name = merchant_name.strip().casefold()
    row = connection.execute(
        """
        INSERT INTO merchants (name, normalized_name, default_category_id)
        VALUES (%s, %s, %s)
        ON CONFLICT (normalized_name)
        DO UPDATE SET default_category_id = EXCLUDED.default_category_id, updated_at = now()
        RETURNING id
        """,
        (merchant_name.strip(), normalized_name, category_id),
    ).fetchone()
    if row is None:
        raise RuntimeError("merchant upsert did not return an id")
    return _read_int(row[0])


def _category_id(connection: Connection[tuple[object, ...]], category_name: str) -> int:
    row = connection.execute("SELECT id FROM categories WHERE name = %s", (category_name,)).fetchone()
    if row is None:
        raise ValueError(f"unknown category: {category_name}")
    return _read_int(row[0])


def _parse_priority(value: str) -> int:
    try:
        priority = int(value or "100")
    except ValueError as exc:
        raise ValueError("rule priority must be an integer") from exc
    if priority < 0:
        raise ValueError("rule priority must be non-negative")
    return priority


def _optional_str(value: object) -> str | None:
    if value is None:
        return None
    return str(value)


def _optional_bool(value: object) -> bool | None:
    if value is None:
        return None
    if not isinstance(value, bool):
        raise RuntimeError(f"expected boolean, got {type(value).__name__}")
    return value


def _read_decimal(value: object) -> Decimal:
    if not isinstance(value, Decimal):
        raise RuntimeError(f"expected decimal, got {type(value).__name__}")
    return value


def _read_int(value: object) -> int:
    if not isinstance(value, int):
        raise RuntimeError(f"expected integer, got {type(value).__name__}")
    return value

File: src/server/management/rules.ts
import type pg from "pg";

import type { RuleEditorRequest, RulePreviewRequest } from "../../shared/management.ts";
import { sql, toQuery } from "../db/sql.ts";
import { listCategories } from "./service.ts";
import { type TransactionMatchRow, transactionRow } from "./transactions.ts";

const ruleFields = [
  "account_id",
  "amount",
  "counterparty_iban",
  "counterparty_name",
  "description",
  "merchant",
] as const;
const ruleOperators = [
  "amount_between",
  "contains",
  "ends_with",
  "exact",
  "regex",
  "starts_with",
] as const;

type RuleRow = {
  category_id: string | null;
  category_name: string | null;
  field: (typeof ruleFields)[number];
  id: string;
  is_active: boolean;
  merchant_name: string | null;
  name: string;
  operator: (typeof ruleOperators)[number];
  pattern: string;
  priority: number;
  set_is_excluded_from_budget: boolean | null;
  set_is_fixed_cost: boolean | null;
  set_is_income: boolean | null;
  set_is_savings: boolean | null;
  set_is_transfer: boolean | null;
};

export async function listRules(pool: pg.Pool) {
  const categories = await listCategories(pool);
  const rows = await pool.query<RuleRow>(`
    SELECT
      categorization_rules.id,
      categorization_rules.name,
      categorization_rules.priority,
      categorization_rules.is_active,
      categorization_rules.field,
      categorization_rules.operator,
      categorization_rules.pattern,
      categories.id::text AS category_id,
      categories.name AS category_name,
      merchants.name AS merchant_name,
      categorization_rules.set_is_income,
      categorization_rules.set_is_transfer,
      categorization_rules.set_is_savings,
      categorization_rules.set_is_fixed_cost,
      categorization_rules.set_is_excluded_from_budget
    FROM categorization_rules
    LEFT JOIN categories ON categories.id = categorization_rules.category_id
    LEFT JOIN merchants ON merchants.id = categorization_rules.merchant_id
    ORDER BY categorization_rules.priority, categorization_rules.id
  `);
  const rules = await Promise.all(
    rows.rows.map(async (row) => {
      const preview = await previewStoredRule(pool, Number(row.id));
      return {
        alreadyCorrectCount: preview.alreadyCorrectCount,
        categoryId: row.category_id ? Number(row.category_id) : null,
        categoryName: row.category_name,
        field: row.field,
        flags: flags(row),
        id: Number(row.id),
        isActive: row.is_active,
        manualOverridesSkippedCount: preview.skippedManualCount,
        matchCount: preview.matchCount,
        merchantName: row.merchant_name,
        name: row.name,
        operator: row.operator,
        pattern: row.pattern,
        priority: row.priority,
        wouldChangeCount: preview.wouldChangeCount,
      };
    }),
  );
  return { categories, fields: ruleFields, operators: ruleOperators, rules };
}

export async function previewRule(pool: pg.Pool, rule: RulePreviewRequest) {
  const result = await matchingTransactions(pool, rule);
  return previewFromMatches(rule, result.rows);
}

export async function createRule(pool: pg.Pool, rule: RuleEditorRequest): Promise<number> {
  const merchantId = await upsertMerchant(pool, rule.merchantName, rule.categoryId);
  const result = await pool.query<{ id: string }>(
    toQuery(sql`
      INSERT INTO categorization_rules (
        name, priority, is_active, field, operator, pattern, category_id, merchant_id,
        set_is_income, set_is_transfer, set_is_savings, set_is_fixed_cost,
        set_is_excluded_from_budget
      )
      VALUES (
        ${rule.name}, ${rule.priority}, ${rule.isActive}, ${rule.field}, ${rule.operator},
        ${rule.pattern}, ${rule.categoryId}, ${merchantId}, ${flagValue(rule.flags.setIsIncome)},
        ${flagValue(rule.flags.setIsTransfer)}, ${flagValue(rule.flags.setIsSavings)},
        ${flagValue(rule.flags.setIsFixedCost)}, ${flagValue(rule.flags.setIsExcludedFromBudget)}
      )
      RETURNING id
    `),
  );
  return Number(result.rows[0]?.id);
}

export async function updateRule(
  pool: pg.Pool,
  ruleId: number,
  rule: RuleEditorRequest,
): Promise<void> {
  const merchantId = await upsertMerchant(pool, rule.merchantName, rule.categoryId);
  await pool.query(
    toQuery(sql`
      UPDATE categorization_rules
      SET
        name = ${rule.name},
        priority = ${rule.priority},
        is_active = ${rule.isActive},
        field = ${rule.field},
        operator = ${rule.operator},
        pattern = ${rule.pattern},
        category_id = ${rule.categoryId},
        merchant_id = ${merchantId},
        set_is_income = ${flagValue(rule.flags.setIsIncome)},
        set_is_transfer = ${flagValue(rule.flags.setIsTransfer)},
        set_is_savings = ${flagValue(rule.flags.setIsSavings)},
        set_is_fixed_cost = ${flagValue(rule.flags.setIsFixedCost)},
        set_is_excluded_from_budget = ${flagValue(rule.flags.setIsExcludedFromBudget)},
        updated_at = now()
      WHERE id = ${ruleId}
    `),
  );
}

export async function setRuleActive(
  pool: pg.Pool,
  ruleId: number,
  isActive: boolean,
): Promise<void> {
  await pool.query(
    toQuery(sql`
      UPDATE categorization_rules
      SET is_active = ${isActive}, updated_at = now()
      WHERE id = ${ruleId}
    `),
  );
}

export async function applyRule(pool: pg.Pool, ruleId: number) {
  const rule = await readRule(pool, ruleId);
  if (!rule.isActive) {
    return { skippedManualCount: 0, updatedCount: 0 };
  }
  const preview = await previewRule(pool, rule);
  if (preview.matches.length === 0) {
    return { skippedManualCount: preview.skippedManualCount, updatedCount: 0 };
  }
  await pool.query(
    `
      UPDATE enriched_transactions
      SET category_id = COALESCE($2::bigint, category_id),
          merchant_id = COALESCE($3::bigint, merchant_id),
          is_income = COALESCE($4::boolean, is_income),
          is_transfer = COALESCE($5::boolean, is_transfer),
          is_savings = COALESCE($6::boolean, is_savings),
          is_fixed_cost = COALESCE($7::boolean, is_fixed_cost),
          is_excluded_from_budget = COALESCE($8::boolean, is_excluded_from_budget),
          needs_review = false,
          classification_method = 'rule',
          classification_confidence = 1,
          classification_reason = $9,
          rule_id = $10,
          updated_at = now()
      WHERE id = ANY($1::bigint[])
        AND NOT EXISTS (
          SELECT 1 FROM manual_overrides
          WHERE manual_overrides.enriched_transaction_id = enriched_transactions.id
        )
    `,
    [
      preview.matches.map((match) => match.id),
      rule.categoryId,
      await merchantIdByName(pool, rule.merchantName),
      flagValue(rule.flags.setIsIncome),
      flagValue(rule.flags.setIsTransfer),
      flagValue(rule.flags.setIsSavings),
      flagValue(rule.flags.setIsFixedCost),
      flagValue(rule.flags.setIsExcludedFromBudget),
      `Backfilled from rule: ${rule.name}`,
      ruleId,
    ],
  );
  return { skippedManualCount: preview.skippedManualCount, updatedCount: preview.matches.length };
}

async function previewStoredRule(pool: pg.Pool, ruleId: number) {
  const rule = await readRule(pool, ruleId);
  const result = await matchingTransactions(pool, rule);
  return previewFromMatches(rule, result.rows);
}

async function readRule(
  pool: pg.Pool,
  ruleId: number,
): Promise<RulePreviewRequest & { isActive: boolean }> {
  const result = await pool.query<RuleRow>(
    `
      SELECT
        categorization_rules.id, categorization_rules.name, categorization_rules.priority,
        categorization_rules.is_active, categorization_rules.field, categorization_rules.operator,
        categorization_rules.pattern, categories.id::text AS category_id, categories.name AS category_name,
        merchants.name AS merchant_name, categorization_rules.set_is_income,
        categorization_rules.set_is_transfer, categorization_rules.set_is_savings,
        categorization_rules.set_is_fixed_cost, categorization_rules.set_is_excluded_from_budget
      FROM categorization_rules
      LEFT JOIN categories ON categories.id = categorization_rules.category_id
      LEFT JOIN merchants ON merchants.id = categorization_rules.merchant_id
      WHERE categorization_rules.id = $1
    `,
    [ruleId],
  );
  const row = result.rows[0];
  if (!row?.category_id) {
    throw new Error("Rule not found");
  }
  return {
    categoryId: Number(row.category_id),
    field: row.field,
    flags: {
      setIsExcludedFromBudget: row.set_is_excluded_from_budget === true,
      setIsFixedCost: row.set_is_fixed_cost === true,
      setIsIncome: row.set_is_income === true,
      setIsSavings: row.set_is_savings === true,
      setIsTransfer: row.set_is_transfer === true,
    },
    isActive: row.is_active,
    merchantName: row.merchant_name ?? undefined,
    name: row.name,
    operator: row.operator,
    pattern: row.pattern,
    priority: row.priority,
  };
}

function matchingTransactions(pool: pg.Pool, rule: RulePreviewRequest) {
  return pool.query<TransactionMatchRow>(
    `
      SELECT enriched_transactions.id, raw_transactions.booking_date::text,
        raw_transactions.amount::text, COALESCE(merchants.name, raw_transactions.counterparty_name, 'Unknown') AS merchant,
        raw_transactions.description, categories.id::text AS category_id, categories.name AS category_name,
        COALESCE(enriched_transactions.classification_method, 'none') AS classification_method,
        enriched_transactions.needs_review, enriched_transactions.is_income, enriched_transactions.is_transfer,
        enriched_transactions.is_savings, enriched_transactions.is_fixed_cost, enriched_transactions.is_recurring,
        enriched_transactions.is_one_off, enriched_transactions.is_excluded_from_budget,
        manual_overrides.id IS NOT NULL AS has_manual_override,
        categories.id::bigint = $4::bigint AS category_matches,
        COALESCE(merchants.name, '') = COALESCE($5::text, merchants.name, '') AS merchant_matches
      FROM enriched_transactions
      JOIN raw_transactions ON raw_transactions.id = enriched_transactions.raw_transaction_id
      LEFT JOIN merchants ON merchants.id = enriched_transactions.merchant_id
      LEFT JOIN categories ON categories.id = enriched_transactions.category_id
      LEFT JOIN manual_overrides ON manual_overrides.enriched_transaction_id = enriched_transactions.id
      WHERE (${matchCondition(rule)})
        AND ($2::numeric IS NULL OR true)
        AND ($3::numeric IS NULL OR true)
      ORDER BY raw_transactions.booking_date DESC, enriched_transactions.id DESC
    `,
    [
      rule.pattern,
      amountLower(rule.pattern),
      amountUpper(rule.pattern),
      rule.categoryId,
      rule.merchantName ?? null,
    ],
  );
}

function matchCondition(rule: RulePreviewRequest): string {
  if (rule.operator === "amount_between") {
    return rule.field === "amount" ? "raw_transactions.amount BETWEEN $2 AND $3" : "FALSE";
  }
  const field = fieldExpression(rule.field);
  switch (rule.operator) {
    case "contains":
      return `${field} ILIKE '%' || $1 || '%'`;
    case "ends_with":
      return `${field} ILIKE '%' || $1`;
    case "exact":
      return `lower(${field}) = lower($1)`;
    case "regex":
      return `${field} ~* $1`;
    case "starts_with":
      return `${field} ILIKE $1 || '%'`;
    default:
      return "FALSE";
  }
}

function fieldExpression(field: RulePreviewRequest["field"]): string {
  switch (field) {
    case "account_id":
      return "raw_transactions.account_id::text";
    case "counterparty_iban":
      return "COALESCE(raw_transactions.counterparty_iban, '')";
    case "counterparty_name":
      return "COALESCE(raw_transactions.counterparty_name, '')";
    case "description":
      return "raw_transactions.description";
    case "merchant":
      return "COALESCE(merchants.name, '')";
    case "amount":
      return "raw_transactions.amount::text";
  }
}

function amountLower(pattern: string): number | null {
  const [lower] = pattern.split("..");
  const parsed = Number(lower);
  return Number.isFinite(parsed) ? parsed : null;
}

function amountUpper(pattern: string): number | null {
  const upper = pattern.split("..")[1];
  const parsed = Number(upper);
  return Number.isFinite(parsed) ? parsed : null;
}

function previewFromMatches(rule: RulePreviewRequest, rows: TransactionMatchRow[]) {
  const skipped = rows.filter((row) => row.has_manual_override).length;
  const eligible = rows.filter((row) => !row.has_manual_override);
  const alreadyCorrect = eligible.filter((row) => alreadyCorrectForRule(rule, row)).length;
  const matches = eligible.filter((row) => !alreadyCorrectForRule(rule, row)).map(transactionRow);
  return {
    alreadyCorrectCount: alreadyCorrect,
    matchCount: rows.length,
    matches,
    skippedManualCount: skipped,
    wouldChangeCount: matches.length,
  };
}

function alreadyCorrectForRule(rule: RulePreviewRequest, row: TransactionMatchRow): boolean {
  return (
    row.category_matches &&
    row.merchant_matches &&
    (!rule.flags.setIsIncome || row.is_income) &&
    (!rule.flags.setIsTransfer || row.is_transfer) &&
    (!rule.flags.setIsSavings || row.is_savings) &&
    (!rule.flags.setIsFixedCost || row.is_fixed_cost) &&
    (!rule.flags.setIsExcludedFromBudget || row.is_excluded_from_budget)
  );
}

async function upsertMerchant(
  pool: pg.Pool,
  merchantName: string | undefined,
  categoryId: number,
): Promise<number | null> {
  const cleaned = merchantName?.trim();
  if (!cleaned) {
    return null;
  }
  const result = await pool.query<{ id: string }>(
    toQuery(sql`
      INSERT INTO merchants (name, normalized_name, default_category_id)
      VALUES (${cleaned}, ${cleaned.toLowerCase()}, ${categoryId})
      ON CONFLICT (normalized_name)
      DO UPDATE SET default_category_id = EXCLUDED.default_category_id, updated_at = now()
      RETURNING id
    `),
  );
  return Number(result.rows[0]?.id);
}

async function merchantIdByName(
  pool: pg.Pool,
  merchantName: string | undefined,
): Promise<number | null> {
  if (!merchantName?.trim()) {
    return null;
  }
  const result = await pool.query<{ id: string }>("SELECT id FROM merchants WHERE name = $1", [
    merchantName.trim(),
  ]);
  return result.rows[0] ? Number(result.rows[0].id) : null;
}

function flagValue(enabled: boolean): boolean | null {
  return enabled ? true : null;
}

function flags(row: RuleRow): string[] {
  return [
    row.set_is_income ? "income" : null,
    row.set_is_transfer ? "transfer" : null,
    row.set_is_savings ? "savings" : null,
    row.set_is_fixed_cost ? "fixed" : null,
    row.set_is_excluded_from_budget ? "excluded" : null,
  ].filter((flag): flag is string => flag !== null);
}
